### PR TITLE
Add new Xtext editor quickfixes on *.xtext files

### PR DESCRIPTION
**Pull request by nbhusare**
- [x] Provide quick-fixes for illegal rule references (see #331943)
- [x] Provide quick-fixes for illegal hidden token definitions (see #331942)
- [x] Provide more quick-fixes for missing rules (see #331941) 

Signed-off-by: nbhusare <neerajbhusare@gmail.com>

### CODE REVIEW

File: org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/editor/quickfix/XtextGrammarQuickfixProvider.java [hunk 1]
```diff
@@ -155,6 +158,39 @@ public void apply(final EObject element, IModificationContext context) throws Ba
 		createLinkingIssueResolutions(issue, acceptor);
 	}
 
+	@Fix(XtextLinkingDiagnosticMessageProvider.UNRESOLVED_TERMINAL_RULE)
+	public void fixUnresolvedTerminalRule(final Issue issue, IssueResolutionAcceptor acceptor) {
+		final String ruleName = issue.getData()[0];
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/editor/quickfix/XtextGrammarQuickfixProvider.java:163]: Could you please extract the common parts into  a single method? fixUnresolvedTerminalRule and fixUnresolvedTerminalFragmentRule are almost the same.

File: org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/editor/quickfix/XtextGrammarQuickfixProvider.java [hunk 2]
```diff
@@ -600,4 +636,64 @@ private void applyToNode(AbstractElement node, String actionText, MultiTextEdit
...
+					IXtextDocument xtextDocument = context.getXtextDocument();
+					xtextDocument.modify(new IUnitOfWork.Void<XtextResource>() {
+
+						@Override
+						public void process(XtextResource state) throws Exception {
+							if (state == null) {
+								return;
+							}
+
+							XtextGrammarAccess grammarAccess = state.getResourceServiceProvider().get(XtextGrammarAccess.class);
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/editor/quickfix/XtextGrammarQuickfixProvider.java:655]: You can inject the XtextGrammarAccess into the QuickfixProvider

File: org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/editor/quickfix/XtextGrammarQuickfixProvider.java [hunk 3]
```diff
@@ -600,4 +636,64 @@ private void applyToNode(AbstractElement node, String actionText, MultiTextEdit
...
+							RuleCall ruleCall = (RuleCall) state.getEObject(issue.getUriToProblem().fragment());
+							ICompositeNode ruleNode = NodeModelUtils.findActualNodeFor(ruleCall.getRule());
+							for (INode node : ruleNode.getAsTreeIterable()) {
+								if (node.getGrammarElement() != null && node.getGrammarElement() == grammarAccess.getTerminalRuleAccess()
+										.getFragmentFragmentKeyword_2_0_0_0()) {
+									Region fragmentRegion = new Region(node.getOffset(), node.getLength());
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/editor/quickfix/XtextGrammarQuickfixProvider.java:661]: node.getTextRegion() already provides a region. No need to create a JFace region here.

File: org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/editor/quickfix/XtextGrammarQuickfixProvider.java [hunk 4]
```diff
@@ -600,4 +636,64 @@ private void applyToNode(AbstractElement node, String actionText, MultiTextEdit
...
+	}
+
+	@Fix(XtextConfigurableIssueCodes.INVALID_HIDDEN_TOKEN)
+	public void fixInvalidHiddenToken(Issue issue, IssueResolutionAcceptor acceptor) {
+		acceptor.accept(issue, //
+				"Remove hidden token definition", // 
+				"Remove hidden token definition", //
+				NULL_QUICKFIX_IMAGE, //
+				(element, context) -> {
+					ParserRule parserRule = (ParserRule) element;
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/editor/quickfix/XtextGrammarQuickfixProvider.java:692]: Hidden tokens can also be defined on the grammar itself.

File: org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/editor/quickfix/XtextGrammarQuickfixProvider.java [hunk 5]
```diff
@@ -600,4 +636,64 @@ private void applyToNode(AbstractElement node, String actionText, MultiTextEdit
...
+								}
+							}
+						}
+					});
+				});
+	}
+
+	@Fix(XtextConfigurableIssueCodes.INVALID_HIDDEN_TOKEN_FRAGMENT)
+	public void fixInvalidHiddenTokenFragment(Issue issue, IssueResolutionAcceptor acceptor) {
+		acceptor.accept(issue, //
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/editor/quickfix/XtextGrammarQuickfixProvider.java:673]: Please extract the common parts into an own method.
I'm not terribly sure, but I think you can put multiple `@Fix` on the same method.

**Comment by cdietrich** [comment]
@szarnekow @nbhusare what is the current state of this PR?

**Comment by nbhusare** [comment]
@cdietrich I'll push the updated today.

**Comment by nbhusare** [comment]
@cdietrich Replacing this PR with https://github.com/eclipse/xtext-eclipse/pull/1487

**Pull request closed without merging**